### ISSUE TITLE
Is there anyone successfully trained the code with any dataset?

### ISSUE THREAD

**Issue by lazerliu** [zzh8829/yolov3-tf2#107]
I want to validate the code by training VOC2007 and then  write code with half-copy ,but my model can detect nothing,what's on earth the reason?
①use TF Object Detection API to create tf record
②train commond line is "python train.py --batch_size 8 --dataset xxx --val_dataset xxx --mode eager_tf --num_classes 20"
③detect cmd line is "python detect.py --weights ./xxx.tf"

**Comment by NiklasWilson** [comment]
This might not be your problem but the detect.py script takes a --num_classes argument as well, if you exclude it, it defaults to 80. It also accepts a --classes argument to specify your list of classes.

**Comment by lazerliu** [comment]
@NiklasWilson thks,I'll try it.

**Comment by NiklasWilson** [comment]
@lazerliu Let me know if you figure out a solution. Ive finally got training working but my modal also can't detect anything, even on the images used to train it...

**Comment by lazerliu** [comment]
@NiklasWilson I deleted the weights trained before and when I trained again with '--num_classes 20',I got loss nan.I dont know the meaningful loss value,but I trained the official Darknet,loss is 0.0xxx.

**Comment by lazerliu** [comment]
> I want to validate the code by training VOC2007 and then write code with half-copy ,but my model can detect nothing,what's on earth the reason?
> ①use TF Object Detection API to create tf record
> ②train commond line is "python train.py --batch_size 8 --dataset xxx --val_dataset xxx --mode eager_tf --num_classes 20"
> ③detect cmd line is "python detect.py --weights ./xxx.tf"

sorry,,②train commond line is `--num_calsses 80`,as I trained with `--num_calsses 20`got loss nan.....

**Comment by mmaaz60** [comment]
> @lazerliu Let me know if you figure out a solution. Ive finally got training working but my modal also can't detect anything, even on the images used to train it...
> 
> What are your loss/val numbers? And what do they need to be at a minimum to see even inaccurate detections?
> 
> One of my attempts I got
> 
> > train: 78.9618148803711, val: 87.8269271850586
> 
> Which I assume are good? or at least decent enough to show something on images used to test? Im not sure what these numbers represent as they can't be percentages since they start in the 10s of thousands.

Hi @NiklasWilson ,

After training my model when I load the checkpoints using detect.py, I got warnings that all the weights in the checkpoints are not being loaded into the network. Do you get any of these warnings while loading trained model? Is there any work around?
Thanks
![WarningWhileLoadingModelFromCheckpointsAfterTraining](https://user-images.githubusercontent.com/55104795/69498344-d8d1bb00-0f08-11ea-944e-2bde621c2480.png)

**Comment by NiklasWilson** [comment]
@mmaaz60 Yes I have the same problem and have not found a work around.
I switched to using an different repo and have run into none of these problems using it.
If you do find a solution please post it here.

**Comment by Robin2091** [comment]
@NiklasWilson Hi, what repo did you switch to? Can you train yolov3 tiny with tfrecords in that repo?

**Comment by lazerliu** [comment]
> @NiklasWilson I deleted the weights trained before and when I trained again with '--num_classes 20',I got loss nan.I dont know the meaningful loss value,but I trained the official Darknet,loss is 0.0xxx.

The loss nan error I have solved it by adding  `--classes ./data/voc.names`,however,the train is still fault,with train loss 38 and cant detect anything!

**Comment by NiklasWilson** [comment]
@Robin2091 https://github.com/AntonMu/TrainYourOwnYOLO

@lazerliu Ive actually never seen a "nan" error. I provided with a txt that has 2 classes but the resulting moda/weights fro mthis script also does not detect anything for me.

**Comment by PieroCV** [comment]
The main problem is that you cannot change your number of classes. I really tried for several days to train a simple model with 1 and 2 classes and all my attempts failed. So, to solve it, just don't touch the code and train my tfrecords with 80 classes and fine_tune mode.
In addition, you must verify your ".names" file. If your file has names other than your tags (from VoTT, for example), you will get nan values. In addition, for the particular case of VoTT (and some Tfrecord encodings), there are some "corrupt" tfrecord files. A corrupt file causes your entire dataset to fail. To solve this, just try training your model with one by one tfrecord file. You'll see that some tfrecords will train normally and some will get nans.

**Comment by lazerliu** [comment]
@PieroCV just `python train.py xxxxxx --num_classes 80 --transfer fine_tune`?

**Comment by lazerliu** [comment]
@NiklasWilson I tried the `python detect.py xxx --num_classes 20`,it still doesnt work.

**Comment by antongisli** [comment]
I’m a little bit suspicious about these lines in train.py:

`dataset.transform_targets(y, anchors, anchor_masks, 80)))`

If you check this function transform_targets in dataset.py then you will see it is defined like this:

`def transform_targets(y_train, anchors, anchor_masks, classes)`

I haven’t dug more into exactly what this function does (there are no comments :( ), but I’m going to try modifying train.py to pass the num_classes variable to this function instead. 

The issue I have right now is if I use anything value except default (80) classes, I get mismatching shape exceptions. It would make sense if that’s getting thrown due to this mismatch in this function (hard coded 80 vs. what I set). I’ll post back if it helps.

**Comment by Robin2091** [comment]
Hi @antongisli ,

I am training with one class and I don't get any mismatch error when I train. Can you also update me on this problem?

Thanks,
Robin

**Comment by NiklasWilson** [comment]
@antongisli 
The mismatch shape has to do with image sizes. If your images must be 416x416. Supposedly you can pass --size to change this but either way your images will need to be square and I'm not 100% sure 100% of the code base supports other sizes. None the less changing all my images to 416x416 made the shape errors go away.

@lazerliu Not sure what to say about that, did you pass --class ./yourClasses.txt and --num_classes 20?

**Comment by PieroCV** [comment]
> I’m a little bit suspicious about these lines in train.py:
> 
> `dataset.transform_targets(y, anchors, anchor_masks, 80)))`
> 
> If you check this function transform_targets in dataset.py then you will see it is defined like this:
> 
> `def transform_targets(y_train, anchors, anchor_masks, classes)`
> 
> I haven’t dug more into exactly what this function does (there are no comments :( ), but I’m going to try modifying train.py to pass the num_classes variable to this function instead.
> 
> The issue I have right now is if I use anything value except default (80) classes, I get mismatching shape exceptions. It would make sense if that’s getting thrown due to this mismatch in this function (hard coded 80 vs. what I set). I’ll post back if it helps.

Dude, i just try everything about change classes in code. Nothing works. Just use sparse categorical crossentropy as class_loss and use 80 classes even if you have only a single class. It worked for me.

**Comment by mmaaz60** [comment]
Hi All,

I am able to train 20 calss tiny yolov3 using this repo. I just passes --classes <the path to my dataset classes file> --class_num 20 to the train.py script. And my trained model is detecting the objects as well. Not much accurate though. Thanks

**Comment by NiklasWilson** [comment]
> Hi All,
> 
> I am able to train 20 calss tiny yolov3 using this repo. I just passes --classes --class_num 20 to the train.py script. And my trained model is detecting the objects as well. Not much accurate though. Thanks

Thats because tiny-yolo is a lot less accurate, that is why it is 4x faster than non tiny, could also be your training data, make sure to add images with zero annotations to help remove false positives.

Did you have to modify the code at all to train it? or did it just work for you?

**Comment by mmaaz60** [comment]
> @mmaaz60
> 
> > I am able to train 20 calss tiny yolov3 using this repo. I just passes --classes --class_num 20 to the train.py script. And my trained model is detecting the objects as well. Not much accurate though. Thanks
> 
> Thats because tiny-yolo is a lot less accurate, that is why it is 4x faster than non tiny, could also be your training data, make sure to add images with zero annotations to help remove false positives.
> 
> Did you have to modify the code at all to train it? or did it just work for you?

Hi @NiklasWilson ,

I did not modify any code except repeating the dataset and passing the steps_per_epoch and validation_steps to the model.fit() call. Well I am able to successfully trained the network but when loading the trained network getting warning that all the parameters are not being loaded. 

I noticed in the model class that the author is returning different architectures for training and inference. This may be the reason of this warning. What do you think? Thanks

**Comment by NiklasWilson** [comment]
@mmaaz60 Maybe, I have no idea, I ran into the loading errors as well.

**Comment by AnaRhisT94** [comment]
> I want to validate the code by training VOC2007 and then write code with half-copy ,but my model can detect nothing,what's on earth the reason?
> ①use TF Object Detection API to create tf record
> ②train commond line is "python train.py --batch_size 8 --dataset xxx --val_dataset xxx --mode eager_tf --num_classes 20"
> ③detect cmd line is "python detect.py --weights ./xxx.tf"

Check in `models.py` your `yolo_score_threshold`, try to make it lower value. 

> flags.DEFINE_float('yolo_score_threshold', 0.07, 'score threshold')

**Comment by AnaRhisT94** [comment]
> Hi All,
> 
> I am able to train 20 calss tiny yolov3 using this repo. I just passes --classes --class_num 20 to the train.py script. And my trained model is detecting the objects as well. Not much accurate though. Thanks

Check in `models.py` your `yolo_score_threshold`, try to make it lower value.

**Comment by AnaRhisT94** [comment]
> @antongisli
> The mismatch shape has to do with image sizes. Your images must be 416x416. Supposedly you can pass --size to change this but either way your images will need to be square and I'm not sure if 100% of the code base supports other sizes. None the less changing all my images to 416x416 made the shape errors go away.
> 
> @lazerliu Not sure what to say about that, did you pass --class ./yourClasses.txt and --num_classes 20?

The code supports any size of images, when training/detecting, it will just resize to 416 and detect on it.

**Comment by AnaRhisT94** [comment]
> @NiklasWilson I deleted the weights trained before and when I trained again with '--num_classes 20',I got loss nan.I dont know the meaningful loss value,but I trained the official Darknet,loss is 0.0xxx.

The loss is `nan` when the annotations are wrong, Darknet's annotations are `x_center, y_center, width, height` and in this repo the annotations are different: `xmin, x_max, y_min, y_max`. Try first training on one image with one annotation, get that to work, and then you'll succeed with more than 1 image. good luck.

**Comment by AnaRhisT94** [comment]
> @lazerliu Let me know if you figure out a solution. Ive finally got training working but my modal also can't detect anything, even on the images used to train it...
> 
> What are your loss/val numbers? And what do they need to be at a minimum to see even inaccurate detections?
> 
> One of my attempts I got
> 
> > train: 78.9618148803711, val: 87.8269271850586
> 
> Which I assume are good? or at least decent enough to show something on images used to test? Im not sure what these numbers represent as they can't be percentages since they start in the 10s of thousands.

This loss is bad. You should try to get loss lower than 6 in order to detect accurately. Also tune up the parameter in `models.py`, the `yolo_score_threshold` to lower value.

**Comment by lazerliu** [comment]
> > @NiklasWilson I deleted the weights trained before and when I trained again with '--num_classes 20',I got loss nan.I dont know the meaningful loss value,but I trained the official Darknet,loss is 0.0xxx.
> 
> The loss is `nan` when the annotations are wrong, Darknet's annotations are `x_center, y_center, width, height` and in this repo the annotations are different: `xmin, ymin, xmax, y_max`. Try first training on one image with one annotation, get that to work, and then you'll succeed with more than 1 image. good luck. (There's already a code ready to check with one picture in this repo) the idea is to get it work with the correct annotations, then you can generalize it to create .tfrecord with a couple of images, see that it works, then createt a big .tfrecord and continue like that.

Thks,good idea to check the dataset!

**Comment by lazerliu** [comment]
Modify the code via #104,and then just train with `python train.py --dataset xxx.record --val_dataset xxx.record --num_classes 80 --weigths xxx.weights --transfer fine_tune --mode fit ` cmd,you may detect something,it works for me.

**Comment by antongisli** [comment]
Hi, I've managed to do this "properly" i think by following a code snippet that the author hinted at in one of these threads. He said to do something like this:

`def main(_argv):
    if FLAGS.tiny:
        model = YoloV3Tiny(FLAGS.size, training=True,
                           classes=FLAGS.num_classes)
        anchors = yolo_tiny_anchors
        anchor_masks = yolo_tiny_anchor_masks
    else:
        model = YoloV3(FLAGS.size, training=True, classes=FLAGS.num_classes)
        model_pretrained = YoloV3(416, training=True, classes=80)
        model_pretrained.load_weights('./checkpoints/yolov3.tf')
        model.get_layer('yolo_darknet').set_weights(model_pretrained.get_layer('yolo_darknet').get_weights())
        anchors = yolo_anchors
        anchor_masks = yolo_anchor_masks`

**Comment by antongisli** [comment]
The author of the original code said to do something like this:

`        model = YoloV3(FLAGS.size, training=True, classes=FLAGS.num_classes)
        model_pretrained = YoloV3(416, training=True, classes=80)
        model_pretrained.load_weights('./checkpoints/yolov3.tf')
        model.get_layer('yolo_darknet').set_weights(model_pretrained.get_layer('yolo_darknet').get_weights())`

The place to put this is in the train.py script, here:
![image](https://user-images.githubusercontent.com/23321000/70379356-d40aff00-192b-11ea-8e86-54a4015f0ff1.png)

Then, remove other mentions of loading the weights (because you already did it in the model_pretrained.load_weights that he added):
![image](https://user-images.githubusercontent.com/23321000/70379372-06b4f780-192c-11ea-8437-bd6bb002335d.png)

I think but not 100% sure, that this copies the yolo_darknet layer to a new model that has been defined using your number of classes (in my case, 81 - since I added one new class), resulting in a model that accepts 81 classes but has all the weights for everything else from the pretrained model (80).

Since you already loaded the weights - you need to comment out later mentions that try to load the weights directly into the model.

this worked for me and I was able to train (add) a new class. I think that if you try to train over top of an existing class then you are mixing the weights of the pretrained model with your new data in the outer layer, which is probably why you get poor accuracy and bad loss. I could be totally wrong about this, and I have yet to get my data in order for my new class to really test it, but so far it seems to be working. My bash history lost my training commands but next time I do it I will report back exactly what I used (num_classes was 81).

**Comment by moloko777** [comment]
> The author of the original code said to do something like this:
> 
> ` model = YoloV3(FLAGS.size, training=True, classes=FLAGS.num_classes) model_pretrained = YoloV3(416, training=True, classes=80) model_pretrained.load_weights('./checkpoints/yolov3.tf') model.get_layer('yolo_darknet').set_weights(model_pretrained.get_layer('yolo_darknet').get_weights())`
> 
> The place to put this is in the train.py script, here:
> ![image](https://user-images.githubusercontent.com/23321000/70379356-d40aff00-192b-11ea-8e86-54a4015f0ff1.png)
> 
> Then, remove other mentions of loading the weights (because you already did it in the model_pretrained.load_weights that he added):
> ![image](https://user-images.githubusercontent.com/23321000/70379372-06b4f780-192c-11ea-8437-bd6bb002335d.png)
> 
> I think but not 100% sure, that this copies the yolo_darknet layer to a new model that has been defined using your number of classes (in my case, 81 - since I added one new class), resulting in a model that accepts 81 classes but has all the weights for everything else from the pretrained model (80).
> 
> Since you already loaded the weights - you need to comment out later mentions that try to load the weights directly into the model.
> 
> this worked for me and I was able to train (add) a new class. I think that if you try to train over top of an existing class then you are mixing the weights of the pretrained model with your new data in the outer layer, which is probably why you get poor accuracy and bad loss. I could be totally wrong about this, and I have yet to get my data in order for my new class to really test it, but so far it seems to be working. My bash history lost my training commands but next time I do it I will report back exactly what I used (num_classes was 81).
> 
> Note: as I am just hacking right now, I hard coded the path to the weights file (in the image), but it should use the parameter given from command line (./checkpoints/yolov3.tf)

This worked like a charm! I've been hacking away at this for a few days now and following these steps have allowed me to at least train.

I also came across the following code "dataset.transform_targets(y, anchors, anchor_masks, 80)))" and can only assume the 80 relates to the original number of classes from the darknet implementation. Changing this to match my custom number of classes does not seem to have caused any problems,  although can't confirm whether it has actually improved anything.

**Comment by zzh8829** [comment]
Here is a complete tutorial on training with VOC2012 dataset
https://github.com/zzh8829/yolov3-tf2/blob/master/docs/training_voc.md

**Comment by mmaaz60** [comment]
Thanks @zzh8829 , it will really help. Just wanted to if the same docs are valid for training of tiny yolo as well?
